### PR TITLE
Orin: UARTI passthrough to Net-VM

### PR DESCRIPTION
**Pull request by jpruiz84**
Having a USB debug port connected to the Net-VM is useful when the SSH connection is not working or correctly configured. This implementation enables access from the Nvidia Orin AGX debug micro USB port (ttyACM1)  to the Net-VM console.

**NOTE**: this implementation only works with Nvidia Jetson Orin AGX, due is the only 
Nvidia Jetson Orin with the UARTI port available.

## Description of changes

The UARTI is mapped as *serial@31d0000* in the device tree information. This UARTI is connected to 
the Nvidia Orin micro USB debugging port (ttyACM1) with a default speed of 115000bps.

The UARTI passthrough is configured in Ghaf for Microvm. 

The UARTI passthrough configuration declaration is available in: 
[UARTI to Net-VM](https://github.com/jpruiz84/ghaf/blob/uarti/modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix) 

This declaration includes:

- Microvm Qemu extra argument to add the 31d0000.serial to the Net-VM.
- Microvm Qemu extra argument to specify a custom device tree (dtb) for the
  Net-VM that includes the 31d0000.serial as a platform device.
- Microvm disable default serial console, to add virtual pci-serial console.
- Binding service (bindSerial31d0000) for the 31d0000.serial in order to 
  bind this device to the VFIO driver to make it available to Microvm.
- Kernel patch to add a custom device tree (dtb) source code for the
  Net-VM.
- Device tree overlay to host device tree to assign an IOMMU to the 31d0000.serial 
  device, and also a dummy driver

Note: due the Linux kernel is not able to use the console in two UART ports
of the same kind, a virtual pci-serial console was used as qemu console output.

Also, a new udev rule was defined to make group kvm all devices that bind to vfio 
in iommu group 59.

    services.udev.extraRules = ''
              # Make group kvm all devices that bind to vfio in iommu group 59
              SUBSYSTEM=="vfio",KERNEL=="59",GROUP="kvm"
            '';

Also, here the *passthroughs.uarti_net_vm.enable* is enabled by default. This flag
enables the UARTI passthrough to he Net-VM. We recommend to have this enabled by 
default, because it is useful to have access to the Net-VM through the debugging
USB port when the ssh connection does not work.

	hardware.nvidia = {
		virtualization.enable = true;
		virtualization.host.bpmp.enable = false;
		passthroughs.host.uarta.enable = false;
		passthroughs.uarti_net_vm.enable = true;
	};

The flag *virtualization.enable* should be enabled, because it is a 
pre-requirement for *passthroughs.uarti_net_vm.enable*.

Jira task: https://ssrc.atlassian.net/browse/SP-3722

## Checklist for things done

- [X] Summary of the proposed changes in the PR description
- [X] More detailed description in the commit message(s)
- [X] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [X] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [X] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [X] PR linked to architecture documentation and requirement(s) (ticket id)
- [X] Test procedure described (or includes tests). Select one or more:
  - [ ] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [] Author has run `nix flake check --accept-flake-config` and it passes
- [] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [X] Author has added reviewers and removed PR draft status

## Testing

Connect the NVIDIA Jetson Orin AGX debug micro USB to your PC 
and open the serial port on ttyACM1 at 115200 bps. 
You can use picocom with the next command:

	picocom -b 115200 /dev/ttyACM1

Press enter and you will see the Net-VM logging

	net-vm login: 

Also, you can reboot your board and see the full Net-VM boot up logs.

### CODE REVIEW

**Review by mikatammi — CHANGES_REQUESTED** [review]
I didn't read the code completely through with thought yet. The Robot Framework test cases fail for both AGX and NX. NetVM does not seem to start

Some clips from the logs:
```
Jan 14 02:18:07 ghaf-host systemd[1]: Starting Bind UARTI to the vfio-platform driver...
Jan 14 02:18:07 ghaf-host bash[2152]: /nix/store/lm10ywzflq9qfhr4fl0zqxrhiksf28ks-bash-5.2-p15/bin/bash: line 1: echo: write error: No such device
Jan 14 02:18:07 ghaf-host systemd[1]: bindSerial31d0000.service: Main process exited, code=exited, status=1/FAILURE
Jan 14 02:18:07 ghaf-host systemd[1]: bindSerial31d0000.service: Failed with result 'exit-code'.
Jan 14 02:18:07 ghaf-host systemd[1]: Failed to start Bind UARTI to the vfio-platform driver.
```

```
Jan 14 02:16:38 ghaf-host kernel: vfio-pci 0001:01:00.0: vfio_ecap_init: hiding ecap 0x1e@0x160
Jan 14 02:16:38 ghaf-host kernel: vfio-platform 31d0000.serial: cannot attach to SMMU 8000000.iommu whilst already attached to domain on SMMU 12000000.iommu
Jan 14 02:16:38 ghaf-host kernel: vfio-platform 31d0000.serial: VFIO: no reset function found!
```

The microvm@net-vm.service stays at started state but it never actually starts to output any useful log.

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  config = lib.mkIf cfg.enable {
+    ghaf.hardware.nvidia.virtualization.enable = true;
+
+    ghaf.virtualization.microvm.netvm.extraModules = [
+      {
+        microvm.qemu.extraArgs = [
+          # Add custom dtb to Net-VM with 31d0000.serial in platform devices
+          "-dtb"
+          "${config.hardware.deviceTree.package}/tegra234-p3701-ghaf-nev-vm.dtb"
```
Comment (mikatammi) [inline modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix:25]: Should this be net-vm instead of nev-vm?

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      }
+      {
+        name = "Add Net-VM device tree with UARTI in platform devices";
+        patch = ./patches/net_vm_dtb_with_uarti.patch;
+      }
+    ];
+
+    systemd.services.bindSerial31d0000 = {
+      description = "Bind UARTI to the vfio-platform driver";
+      wantedBy = ["multi-user.target"];
```
Comment (mikatammi) [inline modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix:49]: Should this actually be service name of the net-vm?

File: targets/nvidia-jetson-orin/default.nix [hunk 3]
```diff
@@ -49,6 +49,11 @@
           ../../modules/virtualization/microvm/microvm-host.nix
           ../../modules/virtualization/microvm/netvm.nix
           ({config, ...}: {
+            services.udev.extraRules = ''
+              # Make microvm the owner of all vfio devices
+              SUBSYSTEM=="vfio",OWNER="microvm",GROUP="kvm"
```
Comment (mikatammi) [inline targets/nvidia-jetson-orin/default.nix:54]: Could we perhaps get little less broad rule, maybe only for the UART device itself?

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch [hunk 4]
```diff
@@ -0,0 +1,421 @@
+diff --git a/nvidia/platform/t23x/concord/kernel-dts/Makefile b/nvidia/platform/t23x/concord/kernel-dts/Makefile
+index 450191e..d07037c 100644
+--- a/nvidia/platform/t23x/concord/kernel-dts/Makefile
++++ b/nvidia/platform/t23x/concord/kernel-dts/Makefile
+@@ -9,6 +9,7 @@ ifneq ($(filter y,$(CONFIG_ARCH_TEGRA_23x_SOC)),)
+ BUILD_ENABLE=y
+ endif
+ 
++dtb-$(BUILD_ENABLE) += tegra234-p3701-ghaf-nev-vm.dtb
```
Comment (mikatammi) [inline modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch:9]: nev-vm

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+ dtb-$(BUILD_ENABLE) += tegra234-p3701-0000-p3737-0000.dtb
+ dtb-$(BUILD_ENABLE) += tegra234-p3701-0000-as-p3701-0004-p3737-0000.dtb
+ dtb-$(BUILD_ENABLE) += tegra234-p3701-0000-as-p3767-0004-p3737-0000.dtb
+diff --git a/nvidia/platform/t23x/concord/kernel-dts/tegra234-p3701-ghaf-nev-vm.dts b/nvidia/platform/t23x/concord/kernel-dts/tegra234-p3701-ghaf-nev-vm.dts
+new file mode 100644
+index 0000000..119a25a
+--- /dev/null
++++ b/nvidia/platform/t23x/concord/kernel-dts/tegra234-p3701-ghaf-nev-vm.dts
```
Comment (mikatammi) [inline modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch:17]: nev-vm

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix (hunk 1 above)
Comment (jpruiz84) [reply modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/default.nix]: The problem with Robot Framework was that it uses the netvm serial console to qemu for testing if the vm boots correctly. Due the console was redirected to the UART passthrough the test failed. Due the Linux kernel is not able to use the console in two UART ports of the same kind, a virtual pci-serial console was used as qemu console output. For this, the regular console was disabled in microvm by a PR that I requested and that it was merged and tested https://github.com/astro/microvm.nix/pull/194

Also, typos in nev->net were fixed.

File: targets/nvidia-jetson-orin/default.nix (hunk 3 above)
Comment (jpruiz84) [reply targets/nvidia-jetson-orin/default.nix]: Fixed with permissions to the IOMMU group

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch (hunk 4 above)
Comment (jpruiz84) [reply modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch]: tixed

File: modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch (hunk 5 above)
Comment (jpruiz84) [reply modules/hardware/nvidia-jetson-orin/virtualization/passthrough/uarti-net-vm/patches/net_vm_dtb_with_uarti.patch]: fixed

**Comment by jpruiz84** [comment]
I fixed all the problems reported and rebased to e43d37fa11f9eb9b8ef6b0dec3df65a17e4c513d

**Review by leivos-unikie — COMMENTED** [review]
- I was not able to repeat the steps in 'Testing the UARTI on Net-VM'

1. Built flash script
`nix build github:jpruiz84/ghaf/uarti#nvidia-jetson-orin-agx-debug-flash-script`

2. Flashed Orin AXG, booted and connected micro usb cable to laptop.

3. I am able to log in to /dev/ttyACM0 but not to /dev/ttyACM1

4. Checked dmesg
[samuli@nixos:~]$ dmesg | grep ttyACM
[13829.603466] cdc_acm 1-5.4:1.1: ttyACM0: USB ACM device
[13829.603775] cdc_acm 1-5.4:1.3: ttyACM1: USB ACM device
[13829.604074] cdc_acm 1-5.4:1.5: ttyACM2: USB ACM device
[13829.604399] cdc_acm 1-5.4:1.7: ttyACM3: USB ACM device


- Checked that ci-test-automation passes, firefox and zathura can be launched from their icons.


- When I am connected to /dev/ttyACM0 during boot this appears to the ghaf-host login line:

<<< Welcome to NixOS 23.11.20231218.3a9928d (aarch64) - ttyTCU0 >>>

Run 'nixos-help' for the NixOS manual.

ghaf-host login: [   12.308038] vfio-platform 31d0000.serial: cannot attach to SMMU 8000000.iommu whilst already attached to domain on SMMU 12000000.iommu

- Login succeeds anyway

**Comment by jpruiz84** [comment]
@leivos-unikie thanks for your testing. We were testing only with *--fast switch* and this bug was hiden in this mode. It was very useful that your team detected this problem now. The problem is that when we flash the full Nvidia eMMC, it flashed also the UEFI with the new device tree with UARTI disabled, then the VM was not able to turn on it. The fix was to use a different device tree for the host and the UEFI. It was testing with flashing process in my end and it worked. I appreciate if you could repeat the test again and mark * Tested on Jetson Orin NX or AGX aarch64* for me.

**Comment by jpruiz84** [comment]
@mikatammi, I have fixed the problems found by the QA team. Also, I have rebased to [d74b690](https://github.com/tiiuae/ghaf/commit/d74b690db11e5e673f3b1f617b1666997feb2206) (06.03.2024). I appreciate if you can review and merge.

**Comment by leivos-unikie** [comment]
Tested on Orin AGX

`nix build github:jpruiz84/ghaf/uarti#nvidia-jetson-orin-agx-debug-flash-script`

- UARTI passthrough to net-vm works. I can log in to net-vm with `sudo minicom -b 115200 -D /dev/ttyACM1`

After hard reset:
```
Jetson UEFI firmware (version 35.3.1 built on 1980-01-01T00:00:00+00:00)
ESC   to enter Setup.
F11   to enter Boot Manager Menu.
Enter to continue boot.
**  WARNING: Test Key is used.  **
......
Could not locate device tree fixup protocol, skipping.
EFI stub: Booting Linux Kernel...
EFI stub: Using DTB from configuration table
EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
EFI stub: Exiting boot services and installing virtual address map...


<<< Welcome to NixOS 23.11pre-git (aarch64) - ttyAMA0 >>>

Run 'nixos-help' for the NixOS manual.

net-vm login: 
```


After `sudo reboot`:

```
[0;ghaf@net-vm: ~ghaf@net-vm:~]$ [   99.407910] reboot: Restarting system


<<< Welcome to NixOS 23.11pre-git (aarch64) - ttyAMA0 >>>

Run 'nixos-help' for the NixOS manual.

net-vm login: 
```

**Comment by leivos-unikie** [comment]
On Orin NX the serial port to ghaf-host is /dev/ttyUSB0. Should UARTI passthrough to net-vm then be /dev/ttyUSB1?

Anyway net-vm is again not booting on Orin NX (cross-compiled). I went through these steps:

- Build qspi flash script for Orin NX
`nix build github:jpruiz84/ghaf/uarti#nvidia-jetson-orin-nx-debug-from-x86_64-flash-qspi`
- Flash qspi with the resulted flash script
- Build ghaf image for Orin NX
`nix build github:jpruiz84/ghaf/uarti#nvidia-jetson-orin-nx-debug-nodemoapps-from-x86_64`
- dd the image on USB SSD and boot Orin NX from that
- Connect to ghaf-host with `sudo minicom -D /dev/ttyUSB0`

```
[0;ghaf@ghaf-host: ~ghaf@ghaf-host:~]$ microvm -l
net-vm: current(nixos-system-net-vm-23.11pre-git), not booted: systemctl start microvm@net-vm.service
```


```
[0;ghaf@ghaf-host: ~ghaf@ghaf-host:~]$ systemctl status microvm@net-vm.service
● microvm@net-vm.service - MicroVM 'net-vm'
     Loaded: loaded (8;;file://ghaf-host/etc/systemd/system/microvm@.service/etc/systemd/system/microvm@.service8;;; static)8;;
    Drop-In: /nix/store/ssinjg1a63146c5ik2byxdw4iyb957bm-system-units/microvm@n>
             └─8;;file://ghaf-host/nix/store/ssinjg1a63146c5ik2byxdw4iyb957bm-system-units/microvm@net-vm.service.d/overrides.confoverrides.conf8;;8;;
     Active: activating (auto-restart) (Result: exit-code) since Fri 2023-07-28>
    Process: 4743 ExecStartPre=/nix/store/s9b1mwvj6sb90hcpfa0vpyknxmvp43yh-unit>
    Process: 4748 ExecStart=/var/lib/microvms/net-vm/current/bin/microvm-run (c>
    Process: 4751 ExecStopPost=/nix/store/n0h5j9iv5727z8q8cq0kj917r26f6vgd-unit>
   Main PID: 4748 (code=exited, status=203/EXEC)
         IP: 0B in, 0B out
        CPU: 22ms
```

**Comment by leivos-unikie** [comment]
I tried also with `nix build --max-jobs 0 github:jpruiz84/ghaf/uarti#packages.aarch64-linux.nvidia-jetson-orin-nx-debug` but it has the same issue: net-vm not booting.

**Review by leivos-unikie — COMMENTED** [review]
Net-vm not booting on Orin NX.

**Comment by jpruiz84** [comment]
Thanks @leivos-unikie for your quick response and testing. Again your tests are helping a lot to evaluate the corner cases of this implementation. I haven't tested in NX, and I noticed that in NX the UARTI is not available. Then, I have disabled the feature for NX. Also, I have updated the documentation and PR message indicating that this feature is available only for AGX. 

I would greatly appreciate it if you could kindly rerun the test on your end. Thanks for your help.

**Review by leivos-unikie — COMMENTED** [review]
It seems that there is net-vm not booting issue still on cross-compiled Orin AGX (I actually had not tested this target before).

[0;ghaf@ghaf-host: ~ghaf@ghaf-host:~]$ microvm -l
net-vm: current(nixos-system-net-vm-23.11pre-git), not booted: systemctl start microvm@net-vm.service

I built the image with
nix build --verbose --show-trace github:jpruiz84/ghaf/uarti#nvidia-jetson-orin-agx-debug-nodemoapps-from-x86_64-flash-script

**Comment by jpruiz84** [comment]
Thanks @leivos-unikie  for the several tests. I haven't tested the cross compilation. I tested today, and I was able to replicate your results. I found that the new version of microvm that I have to use here has a bug with cross compilation. I will fix the bug, made the PR to micro VM, update again here, and let you know for testing again.

**Comment by jpruiz84** [comment]
Hi @leivos-unikie . I have fixed the problems with cross compilation. I appreciate if you could please help me to test again?

**Comment by leivos-unikie** [comment]
> Hi @leivos-unikie . I have fixed the problems with cross compilation. I appreciate if you could please help me to test again?

Tested cross-compiled flash-script for Orin AGX. Now net-vm boots and UARTI passthrough to it works.

**Review by leivos-unikie — APPROVED** [review]

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
+<!--
+    Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+    SPDX-License-Identifier: CC-BY-SA-4.0
+-->
+
+# Device tree overlays for passthrough
+
+The device tree blob (.dtb file) is a data structure that describes the hardware
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:8]: (.dtb file) -> (DTB)

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+-->
+
+# Device tree overlays for passthrough
+
+The device tree blob (.dtb file) is a data structure that describes the hardware 
+components of a particular system so that the operating system can use and manage 
+those components. For passthrough, the hardware description in the host needs
+some modifications, which include:
+
+- Removing the driver of the device to passthrough by assigning a dummy string
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:13]: the driver of the device -> the device's driver

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  to the *compatible* property.
+- Adding the *iommus* property to the device to passthrough.
+- Removing or adding other properties that cause conflicts during the passthrough.
+
+Modifying the host device tree could be done by applying patches
+to the .dts (Device Tree Source) files. Nevertheless, this option is not
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:19]: to the .dts (Device Tree Source) files -> to DTS files

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+some modifications, which include:
+
+- Removing the driver of the device to passthrough by assigning a dummy string
+  to the *compatible* property.
+- Adding the *iommus* property to the device to passthrough.
+- Removing or adding other properties that cause conflicts during the passthrough.
+
+Modifying the host device tree could be done by applying patches
+to the .dts (Device Tree Source) files. Nevertheless, this option is not 
+scalable if we need to apply different patches to the same .dts from
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:20]: .dts -> DTS file

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+  to the *compatible* property.
+- Adding the *iommus* property to the device to passthrough.
+- Removing or adding other properties that cause conflicts during the passthrough.
+
+Modifying the host device tree could be done by applying patches
+to the .dts (Device Tree Source) files. Nevertheless, this option is not 
+scalable if we need to apply different patches to the same .dts from 
+different configurations.
+
+A better and more scalable approach to modifying a device tree is using device
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:23]: A better and more scalable approach **for** modifying a device tree....

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+tree overlays. The device tree overlay contains information about the nodes
+to modify (in nodes called *fragment@0 ...*) and the overlay of the 
+properties that we want to affect. More information about the overlays
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:26]: For more information on the overlays, see [Overlay notes](https://www.kernel.org/doc/Documentation/devicetree/overlay-notes.txt).

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+can be found in [Overlay notes](https://www.kernel.org/doc/Documentation/devicetree/overlay-notes.txt).
+
+
+## The Nix hardware.deviceTree module
+
+The Nix hardware.deviceTree module is a useful tool to handle the device
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:32]: The Nix hardware.deviceTree module helps to work with the device trees and their overlays:

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+trees and their overlays. Here you can:
+
+- Define the device tree overlay file to use.
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:35]: To define the device tree overlay file to use.

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+- Use a filter to apply the overlay only to specific files.
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:36]: To use a filter to apply the overlay only to specific files.

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+- Define the include paths to build the device tree overlay.
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:37]: To define included paths to build the device tree overlay.

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+The next section shows an example of an overlay for the UARTI passthrough
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:39]: Delete from here.

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+## Device tree overlay example
+
```
Comment (jenninikko) [inline docs/src/technologies/device_tree_overlays_pt.md:42]: Add here:

In this section, you can find an example of an overlay for the UARTI passthrough.

**Review by jenninikko — CHANGES_REQUESTED** [review]
As usual, a couple of comments from me :D

**Comment by jpruiz84** [comment]
As always, thanks @jenninikko for your very useful comments. I have fixed all of them. Please help to review.

File: docs/src/technologies/device_tree_overlays_pt.md [hunk 18]
```diff
@@ -0,0 +1,150 @@
+<!--
+    Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+    SPDX-License-Identifier: CC-BY-SA-4.0
+-->
+
+# Device Tree Overlays for Passthrough
+
+The device tree blob (DTB) is a data structure that describes the hardware
```
Comment (mikatammi) [inline docs/src/technologies/device_tree_overlays_pt.md:8]: Again feels like most minor nitpick of all times, but this file has multiple lines with trailing whitespace. Clearly visible with red color in `git diff` from terminal. Could you please remove those?

**Review by mikatammi — APPROVED** [review]
Tested and technically works nicely 👍

**Review by mikatammi — CHANGES_REQUESTED** [review]
Do not merge. This has microvm flake.lock update which breaks everything in Lenovo X1 Carbon

**Comment by jpruiz84** [comment]
Hi @mikatammi .  The PR https://github.com/astro/microvm.nix/pull/216 to opt back for q35 was merged in microvm.nix. Could you please update Ghaf main line guivm definition to use the latest micrivm.nix version? In this way I can rebase my PR to pass all the targets.

**Comment by mikatammi** [comment]
After rebasing on top of this https://github.com/tiiuae/ghaf/pull/561, this UARTI no longer seems to work :(

File: flake.lock
Comment (brianmcgillion) [inline flake.lock:1]: Please drop this flake update

**Comment by jpruiz84** [comment]
Thanks @mikatammi and @brianmcgillion for updating the microvm.nix flake. I rebased droping my flake.look changes, also I fixed my device tree overlay bud with jetpack 35.4.1. Now it is working in my side. @leivos-unikie could you please test in your side, and help me marking the tested checkbox?

**Review by leivos-unikie — APPROVED** [review]
Tests

Orin AGX (native build)
- Checked UARTI passthrough
- ci-test-automation pass
- Apps launch

Orin AGX (cross-compiled)
- Checked UARTI passthrough
- ci-test-automation pass
- Apps launch

Orin NX (cross-compiled)
- Boot test, also net-vm boot ok
- Apps launch

Lenovo-X
- Boot, apps launch

**Review by jenninikko — APPROVED** [review]

**Comment by jpruiz84** [comment]
Hi @mikatammi and @brianmcgillion. All test passed by QA team, also documentation is approved. Could you please approve this PR?

**Review by brianmcgillion — APPROVED** [review]
